### PR TITLE
[1LP][RFR] Add UI tests to test the buttons in the physical server details page

### PR DESCRIPTION
**Pull request by CharlleDaniel**
This PR is able to : 
* Add new tests to cover all options in the `Power` button are working on physical server details page.
* Fix some issues that aren't working more in the `physical_server.py`

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit a23d73024f0ff47a29d4915e04893b0e657c4c3b:

`cfme/physical/physical_server.py`:
- :red_circle: Line 24:1: **F401** *'time' imported but unused*
- :bangbang: Line 127:101: **E501** *line too long (130 > 100 characters)*
- :bangbang: Line 146:101: **E501** *line too long (102 > 100 characters)*

`cfme/tests/physical_infrastructure/ui/test_physical_server_power_actions.py`:
- :red_circle: Line 4:1: **F401** *'cfme.utils.appliance.implementations.ui.navigate_to' imported but unused*
- :red_circle: Line 6:1: **F401** *'cfme.utils.wait.wait_for' imported but unused*
- :red_circle: Line 7:1: **F401** *'cfme.utils.log.logger' imported but unused*
- :bangbang: Line 22:82: **E231** *missing whitespace after ','*
- :bangbang: Line 54:83: **E231** *missing whitespace after ','*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ec275fcb350be827775bcb2d52bf40e3ca2aab07:

`cfme/physical/physical_server.py`:
- :bangbang: Line 120:101: **E501** *line too long (108 > 100 characters)*
- :bangbang: Line 126:101: **E501** *line too long (130 > 100 characters)*
- :bangbang: Line 146:101: **E501** *line too long (105 > 100 characters)*

`cfme/tests/physical_infrastructure/ui/test_physical_server_power_actions.py`:
- :red_circle: Line 4:1: **F401** *'cfme.utils.appliance.implementations.ui.navigate_to' imported but unused*
- :red_circle: Line 6:1: **F401** *'cfme.utils.wait.wait_for' imported but unused*
- :red_circle: Line 7:1: **F401** *'cfme.utils.log.logger' imported but unused*
- :bangbang: Line 22:82: **E231** *missing whitespace after ','*
- :bangbang: Line 54:83: **E231** *missing whitespace after ','*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 428110eba585ea462f372e32d0ffd1bb0812a842:

`cfme/physical/physical_server.py`:
- :bangbang: Line 120:101: **E501** *line too long (108 > 100 characters)*
- :bangbang: Line 146:101: **E501** *line too long (105 > 100 characters)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by CharlleDaniel** [comment]
My PR did not break the CIs, it had failed because of another thing.

File: cfme/tests/physical_infrastructure/ui/test_physical_server_power_actions.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+from cfme.physical.provider.lenovo import LenovoProvider
+
+pytestmark = [pytest.mark.tier(3), pytest.mark.provider([LenovoProvider], scope="module")]
+
+
+@pytest.fixture(scope="module")
+def physical_server(appliance, provider):
+    # Get and return the first physical server
+    ph_server = appliance.collections.physical_servers.all(provider)[0]
+    yield ph_server
```
Comment (quarckster) [inline cfme/tests/physical_infrastructure/ui/test_physical_server_power_actions.py]: **REQUIRED**
```python
return ph_server
```

File: cfme/physical/physical_server.py [hunk 2]
```diff
@@ -105,21 +117,33 @@ def refresh(self, cancel=False):
         view.toolbar.configuration.item_select("Refresh Relationships and Power States",
             handle_alert=cancel)
 
-    def wait_for_physical_server_state_change(self, desired_state, timeout=300):
+    def wait_for_state_change(self, desired_state, target, provider, timeout=300, delay=10):
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
This method should be private and called in other methods such as `power_off` and `power_on`:
```python
def power_on(self, wait=True):
    if wait:
        self._wait_for_state_change(...)
    ...
```

**Comment by dajoRH** [comment]
Lint report for commit e682f334e316e8e6482867b9591320d7edcb651f:

`cfme/physical/physical_server.py`:
- :bangbang: Line 295:101: **E501** *line too long (102 > 100 characters)*
- :bangbang: Line 351:1: **E302** *expected 2 blank lines, found 1*

`cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py`:
- :bangbang: Line 8:1: **E302** *expected 2 blank lines, found 1*
- :bangbang: Line 17:4: **E111** *indentation is not a multiple of four*
- :bangbang: Line 18:4: **E111** *indentation is not a multiple of four*
- :bangbang: Line 19:4: **E111** *indentation is not a multiple of four*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 715d3da6ad0b44525a0d8816df2575240f51f20c:

`cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py`:
- :bangbang: Line 18:4: **E111** *indentation is not a multiple of four*
- :bangbang: Line 19:4: **E111** *indentation is not a multiple of four*
- :bangbang: Line 20:4: **E111** *indentation is not a multiple of four*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by CharlleDaniel** [comment]
@quarckster Thank you for your review! Could you review again?

File: cfme/tests/physical_infrastructure/ui/test_physical_server_power_actions.py (hunk 1 above)
Comment (CharlleDaniel) [reply cfme/tests/physical_infrastructure/ui/test_physical_server_power_actions.py]: done

File: cfme/physical/physical_server.py (hunk 2 above)
Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: done

File: cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+def test_refresh_relationships(physical_server, provider):
+    last_refresh = provider.last_refresh_date()
+    physical_server.refresh(provider, handle_alert=True)
+    assert last_refresh != provider.last_refresh_date()
+
+
+# Power Button
+def test_power_off(physical_server, provider):
+    flash_messages = physical_server.power_off()
+    assert flash_messages[0].text == 'Requested Server power_off for the selected server'
```
Comment (quarckster) [inline cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py]: **REQUIRED**
Here and further is not the best approach. `physical_server.power_off()` should return nothing. You need to create an appropriate view and get flash message from there:
```python
physical_server.power_off()
view = provider.create_view("some_view")
view.flash.assert_message('Requested Server power_off for the selected server')
```

File: cfme/physical/physical_server.py [hunk 4]
```diff
@@ -68,58 +69,90 @@ def load_details(self, refresh=False):
...
+    def execute_button(self, button, option, handle_alert=False):
         view = navigate_to(self, "Details")
-        view.toolbar.custom_button(button_group).item_select(button, handle_alert=handle_alert)
+        view.toolbar.custom_button(button).item_select(option, handle_alert=handle_alert)
+        return view
 
-    def power_on(self):
-        view = navigate_to(self, "Details")
-        view.toolbar.power.item_select("Power On", handle_alert=True)
+    def execute_action_button(self, button, option, **kwargs):
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
Please make these methods private.

File: cfme/physical/physical_server.py [hunk 5]
```diff
@@ -68,58 +69,90 @@ def load_details(self, refresh=False):
...
+        self.execute_button('Configuration', 'Refresh Relationships and Power States', handle_alert)
+        wait_for(
+            lambda: last_refresh != provider.last_refresh_date(),
+            message="Wait for the server to be refreshed...",
+            num_sec=300,
+            delay=5
+        )
+
+    def lifecycle_provision(self):
+        return navigate_to(self, "Provision")
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
This method doesn't make sense. In the test just use `view = navigate_to(self, "Provision")`

File: cfme/physical/physical_server.py (hunk 4 above)
Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: sure, Done.

File: cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py (hunk 3 above)
Comment (CharlleDaniel) [reply cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py]: I understand and I agree with you. Thank you so much for help me.

File: cfme/physical/physical_server.py (hunk 5 above)
Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: I change the body of this method, now make sense for you? It's the same approach than the other methods, as `power_on'.

**Comment by CharlleDaniel** [comment]
@quarckster could you review again?

File: cfme/physical/physical_server.py [hunk 6]
```diff
@@ -68,58 +69,90 @@ def load_details(self, refresh=False):
             view.browser.refresh()
             view.flush_widget_cache()
 
-    def execute_button(self, button_group, button, handle_alert=False):
+    def execute_button(self, button, option, handle_alert=False):
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
This should be a private method too.

File: cfme/physical/physical_server.py [hunk 7]
```diff
@@ -68,58 +69,90 @@ def load_details(self, refresh=False):
...
         view = navigate_to(self, "Details")
-        view.toolbar.custom_button(button_group).item_select(button, handle_alert=handle_alert)
+        view.toolbar.custom_button(button).item_select(option, handle_alert=handle_alert)
+        return view
 
-    def power_on(self):
-        view = navigate_to(self, "Details")
-        view.toolbar.power.item_select("Power On", handle_alert=True)
+    def __execute_action_button(self, button, option, handle_alert=True, **kwargs):
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
Only one underscore is required to designate a method as private.
`_private_method()`

File: cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    last_refresh = provider.last_refresh_date()
+    physical_server.refresh(provider, handle_alert=True)
+    assert last_refresh != provider.last_refresh_date()
+
+
+# Power Button
+def test_power_off(physical_server, provider):
+    physical_server.power_off()
+    view = provider.create_view(PhysicalServerDetailsView, physical_server)
+    assert view.flash.messages[0].text == 'Requested Server power_off for the selected server'
```
Comment (quarckster) [inline cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py]: **REQUIRED**
```python
view.flash.assert_message('Requested Server power_off for the selected server')
```

File: cfme/physical/physical_server.py (hunk 6 above)
Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: ok

File: cfme/physical/physical_server.py (hunk 7 above)
Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: OK :confused:

File: cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py (hunk 8 above)
Comment (CharlleDaniel) [reply cfme/tests/physical_infrastructure/ui/test_physical_server_details_buttons.py]: ok, thanks

**Comment by CharlleDaniel** [comment]
@quarckster could you review again?

File: cfme/physical/physical_server.py [hunk 9]
```diff
@@ -68,58 +69,91 @@ def load_details(self, refresh=False):
...
 
-    def wait_for_physical_server_state_change(self, desired_state, timeout=300):
+    def _wait_for_state_change(self, desired_state, target, provider, view, timeout=300, delay=10):
         """Wait for PhysicalServer to come to desired state. This function waits just the needed amount of
            time thanks to wait_for.
 
         Args:
             desired_state (str): 'on' or 'off'
+            target (Method): The Method that most be used to compare with the desired_state.
+            view (Navigate): The view that most be refreshed to verify if the value was changed.
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
According to the code below `target`is a string, `view` is instance of ":py:class:\`widgetastic.widget.View\`"

File: cfme/physical/physical_server.py [hunk 10]
```diff
@@ -68,58 +69,91 @@ def load_details(self, refresh=False):
...
+            lambda: last_refresh != provider.last_refresh_date(),
+            message="Wait for the server to be refreshed...",
+            num_sec=300,
+            delay=5
+        )
+
+    def lifecycle_provision(self):
         view = navigate_to(self, "Details")
-        view.toolbar.power.item_select("Power Off", handle_alert=True)
+        view.toolbar.lifecycle.item_select("Provision Physical Server")
```
Comment (quarckster) [inline cfme/physical/physical_server.py]: **REQUIRED**
I don't understand why do you need this method at all? You have `Provision` navigation step and it does completely the same.

Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: I prefer use the method to keep a pattern in the tests, so all the tests are invoking an action through the button in the physical server page and if I use directly `navigate_to` in the test file I need to import more libraries and I break the pattern.  The Lifecycle is a button in the physical server page and when I select `Provision Physical Server`  the action is navigate to a new view, the same way occurs when I select the Power ON option inside the Power button. The action is queue a job to power on the server and if I can create a method to power_on, I think that I can create a method to select the Lifecycle button as well.

File: cfme/physical/physical_server.py (hunk 9 above)
Comment (CharlleDaniel) [reply cfme/physical/physical_server.py]: I forgot to change the doc string.

File: cfme/physical/physical_server.py (hunk 10 above)
Comment (quarckster) [reply cfme/physical/physical_server.py]: I don't see any problem with importing `navigate_to`. Please follow DRY principle and remove this method.

**Comment by dajoRH** [comment]
Lint report for commit 04fbf41d629b78a49d2b0fe5777c33f4cb951726:

`cfme/physical/physical_server.py`:
- :bangbang: Line 145:101: **E501** *line too long (101 > 100 characters)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7bcbee517e391cdfdc7152f943f4f6a0b16a4e8b:

`cfme/physical/physical_server.py`:
- :bangbang: Line 141:101: **E501** *line too long (101 > 100 characters)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 28082ccc331d510aa986f1f6e51cf1ce64414ce9:

`cfme/physical/physical_server.py`:
- :bangbang: Line 141:101: **E501** *line too long (101 > 100 characters)*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by CharlleDaniel** [comment]
@quarckster could you review again?

**Pull request merged**